### PR TITLE
VP-5616: Extend Shopping cart with Configured Groups

### PR DESCRIPTION
**Pull request by trueboroda**
Extend shopping cart models.
Implement the logic of group`s adding , changing, removing. 
Include into the group data the info about the product and configured product parts.

### CODE REVIEW

File: VirtoCommerce.Storefront.Model/Cart/Demo/ConfiguredGroup.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        public DateTime CreatedDate { get; set; }
+
+        public DateTime? ModifiedDate { get; set; }
+
+        public string CreatedBy { get; set; }
+
+        public string ModifiedBy { get; set; }
+
+        [JsonRequired]
+        public virtual IList<LineItem> Items { get; set; } = new List<LineItem>();
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront.Model/Cart/Demo/ConfiguredGroup.cs]: It is better to initialize all properties in the one place.
Please move collection initialization to constructor.

File: VirtoCommerce.Storefront.Model/Cart/Demo/ShoppingCart.cs [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 namespace VirtoCommerce.Storefront.Model.Cart
 {
     public partial class ShoppingCart
     {
+        [JsonRequired]
+        public LineItem[] UsualItems
+        {
+            get
+            {
+                var result = Items.Where(x => !ConfiguredGroups.Any(y => y.Items.Contains(x))).ToArray();
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront.Model/Cart/Demo/ShoppingCart.cs:15]: Expression could be simplified 
UsualItems => Items.Where(x => !ConfiguredGroups.Any(y => y.Items.Contains(x))).ToArray();

File: VirtoCommerce.Storefront/Controllers/Api/ApiCartDemoController.cs [hunk 3]
```diff
@@ -28,24 +29,48 @@ public class ApiCartDemoController : StorefrontControllerBase
...
 
-
         // POST: storefrontapi/cart/items/bulk
         [HttpPost("items/bulk")]
         [ValidateAntiForgeryToken]
         public async Task<ActionResult<AddItemsToCartResult>> AddItemsToCart([FromBody] AddCartItem[] items)
         {
+            if (items.IsNullOrEmpty())
+            {
+                throw new ArgumentException("Items should not be null or empty");
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Controllers/Api/ApiCartDemoController.cs]: Atgument**Null**Exception

File: VirtoCommerce.Storefront/Controllers/Api/ApiCartDemoController.cs [hunk 4]
```diff
@@ -28,24 +29,48 @@ public class ApiCartDemoController : StorefrontControllerBase
...
+                                      );
+
+                if (configuredGroup == null)
+                {
+                    configuredGroup = new Model.Cart.Demo.ConfiguredGroup(firstItem.Quantity, WorkContext.CurrentCurrency, configuredProductId);
+                    cart.ConfiguredGroups.Add(configuredGroup);
+                }
+                else
+                {
+                    configuredGroup.Quantity = configuredGroup.Quantity + Math.Max(1, firstItem.Quantity);
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Controllers/Api/ApiCartDemoController.cs]: configuredGroup.Quantity += Math.Max(1, firstItem.Quantity);

File: VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs [hunk 5]
```diff
@@ -496,6 +496,11 @@ public static ShoppingCart ToShoppingCart(this cartDto.ShoppingCart cartDto, Cur
             result.VolumetricWeight = (decimal)(cartDto.VolumetricWeight ?? 0);
             result.Weight = (decimal)(cartDto.Weight ?? 0);
 
+            if (cartDto.ConfiguredGroups != null)
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs]: cartDto.ConfiguredGroups.IsNullOrEmpty

File: VirtoCommerce.Storefront/Domain/Cart/Demo/CartConverter.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
+using System;
+using System.Linq;
+using VirtoCommerce.Storefront.Model.Cart;
+using VirtoCommerce.Storefront.Model.Cart.Demo;
+using VirtoCommerce.Storefront.Model.Common;
+using cartDto = VirtoCommerce.Storefront.AutoRestClients.CartModuleApi.Models;
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/CartConverter.cs]: cartApiDto

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 7]
```diff
@@ -34,18 +34,19 @@ public override Task RemoveItemAsync(string lineItemId)
...
 
-            var configureLineItem = Cart.ConfiguredItems.FirstOrDefault(x => x.ConfiguredLineItem?.ProductId.Equals(lineItemId, StringComparison.InvariantCulture) ?? false);
+            var configuredGroup = Cart.ConfiguredGroups?.FirstOrDefault(x => x.Id.Equals(lineItemId, StringComparison.InvariantCulture));
 
-            if (configureLineItem != null)
+            if (configuredGroup != null)
             {
-                var configurablePieces = Cart.Items.Where(x => x.ConfiguredProductId?.Equals(lineItemId, StringComparison.InvariantCulture) ?? false).ToArray();
+                var groupItems = Cart.Items.Where(x => !string.IsNullOrEmpty(x.ConfiguredGropupId) &&
+                       x.ConfiguredGropupId.Equals(configuredGroup.Id)).ToArray();
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: Use invariant comparing

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 8]
```diff
@@ -55,18 +56,16 @@ public override Task ChangeItemQuantityAsync(ChangeCartItemQty changeItemQty)
...
-                    .Items
-                    .Where(x =>
-                        !string.IsNullOrEmpty(x.ConfiguredProductId) &&
-                        x.ConfiguredProductId.Equals(configuredProduct.ConfiguredLineItem.ProductId)
-                    )
-                )
+                configuredGroup.Quantity = changeItemQty.Quantity;
+
+                var groupItems = Cart.Items.Where(x => !string.IsNullOrEmpty(x.ConfiguredGropupId) &&
+                       x.ConfiguredGropupId.Equals(configuredGroup.Id)).ToArray();
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: Use invariant comparing

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 9]
```diff
@@ -116,14 +115,18 @@ public override async Task<bool> AddItemAsync(AddCartItem addCartItem)
 
         protected override async Task AddLineItemAsync(LineItem lineItem)
         {
-            if (!string.IsNullOrEmpty(lineItem.ConfiguredProductId))
+            var existingLineItem = Cart.Items.FirstOrDefault(li => li.ProductId == lineItem.ProductId && li.ConfiguredGropupId == lineItem.ConfiguredGropupId);
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: Use invariant comparing

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 10]
```diff
@@ -116,14 +115,18 @@ public override async Task<bool> AddItemAsync(AddCartItem addCartItem)
...
+            if (existingLineItem != null)
             {
-                lineItem.Id = null;
-                Cart.Items.Add(lineItem);
+                await ChangeItemQuantityAsync(existingLineItem, existingLineItem.Quantity + Math.Max(1, lineItem.Quantity));
+                await ChangeItemPriceAsync(existingLineItem, new ChangeCartItemPrice() { LineItemId = existingLineItem.Id, NewPrice = lineItem.ListPrice.Amount });
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: new ChangeCartItemPrice {

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs [hunk 11]
```diff
@@ -66,38 +66,30 @@ public override async Task<IPagedList<ShoppingCart>> SearchCartsAsync(CartSearch
...
+            var groupProducts = await _catalogService.GetProductsAsync(groupProductsIds, ItemResponseGroup.None);
 
-                if (configuredItem.ConfiguredLineItem != null)
-                {
-                    configuredItem.ConfiguredLineItem.PlacedPrice = new Money(configuredProductItems.Sum(x => x.PlacedPrice.Amount), currency);
-                    configuredItem.ConfiguredLineItem.ExtendedPrice = new Money(configuredProductItems.Sum(x => x.ExtendedPrice.Amount), currency);
-                }
+            foreach (var group in cart.ConfiguredGroups)
+            {
+                var product = groupProducts.FirstOrDefault(x=>x.Id == group.ProductId);
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs]: Use invariant comparing

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Comment by pushnitsa** [comment]
@trueboroda Please fix sonar issues

File: VirtoCommerce.Storefront.Model/Cart/Demo/LineItem.cs [hunk 12]
```diff
@@ -2,6 +2,6 @@ namespace VirtoCommerce.Storefront.Model.Cart
 {
     public partial class LineItem
     {
-        public string ConfiguredProductId { get; set; }
+        public string ConfiguredGropupId { get; set; }
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront.Model/Cart/Demo/LineItem.cs]: I think there is a typo:  it should be ConfiguredGroupId

File: VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs [hunk 13]
```diff
@@ -682,11 +689,9 @@ public static LineItem ToLineItem(this cartDto.CartLineItem lineItemDto, Currenc
                 Width = (decimal?)lineItemDto.Width,
                 Length = (decimal?)lineItemDto.Length,
                 Height = (decimal?)lineItemDto.Height,
-                ConfiguredProductId = lineItemDto.ConfiguredProductId,
+                ConfiguredGropupId = lineItemDto.ConfiguredGroupId,
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs]: same typo

File: VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs [hunk 14]
```diff
@@ -768,7 +773,7 @@ public static cartDto.CartLineItem ToLineItemDto(this LineItem lineItem)
                 DynamicProperties = lineItem.DynamicProperties.Select(ToCartDynamicPropertyDto).ToList(),
                 VolumetricWeight = (double)(lineItem.VolumetricWeight ?? 0),
 
-                ConfiguredProductId = lineItem.ConfiguredProductId,
+                ConfiguredGroupId = lineItem.ConfiguredGropupId,
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 15]
```diff
@@ -34,18 +34,19 @@ public override Task RemoveItemAsync(string lineItemId)
...
             EnsureCartExists();
 
-            var configureLineItem = Cart.ConfiguredItems.FirstOrDefault(x => x.ConfiguredLineItem?.ProductId.Equals(lineItemId, StringComparison.InvariantCulture) ?? false);
+            var configuredGroup = Cart.ConfiguredGroups?.FirstOrDefault(x => x.Id.Equals(lineItemId, StringComparison.InvariantCulture));
 
-            if (configureLineItem != null)
+            if (configuredGroup != null)
             {
-                var configurablePieces = Cart.Items.Where(x => x.ConfiguredProductId?.Equals(lineItemId, StringComparison.InvariantCulture) ?? false).ToArray();
+                var groupItems = Cart.Items.Where(x => !string.IsNullOrEmpty(x.ConfiguredGropupId) &&
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs (hunk 7 above)
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 16]
```diff
@@ -55,18 +56,16 @@ public override Task ChangeItemQuantityAsync(ChangeCartItemQty changeItemQty)
...
-                foreach (var lineItem in Cart
-                    .Items
-                    .Where(x =>
-                        !string.IsNullOrEmpty(x.ConfiguredProductId) &&
-                        x.ConfiguredProductId.Equals(configuredProduct.ConfiguredLineItem.ProductId)
-                    )
-                )
+                configuredGroup.Quantity = changeItemQty.Quantity;
+
+                var groupItems = Cart.Items.Where(x => !string.IsNullOrEmpty(x.ConfiguredGropupId) &&
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs (hunk 8 above)
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs [hunk 17]
```diff
@@ -83,9 +82,9 @@ public override async Task<bool> AddItemAsync(AddCartItem addCartItem)
 
             if (result.IsValid)
             {
-                var lineItem = addCartItem.Product.ToLineItem(Cart.Language, addCartItem.Quantity);                
+                var lineItem = addCartItem.Product.ToLineItem(Cart.Language, addCartItem.Quantity);
                 lineItem.Product = addCartItem.Product;
-                lineItem.ConfiguredProductId = addCartItem.ConfiguredProductId;
+                lineItem.ConfiguredGropupId = addCartItem.ConfiguredGroupId;
```
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs (hunk 9 above)
Comment (ilyawzrd) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartBuilder.cs]: typo

**Review by ilyawzrd — CHANGES_REQUESTED** [review]

File: VirtoCommerce.Storefront.Model/Cart/Demo/LineItem.cs (hunk 12 above)
Comment (trueboroda) [reply VirtoCommerce.Storefront.Model/Cart/Demo/LineItem.cs]: Thank you.

File: VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs (hunk 5 above)
Comment (asvishnyakov) [reply VirtoCommerce.Storefront/Domain/Cart/CartConverter.cs:499]: It's doing same way as below and above. And I think we already discussed about that

File: VirtoCommerce.Storefront/Domain/Cart/Demo/CartConverter.cs (hunk 6 above)
Comment (asvishnyakov) [reply VirtoCommerce.Storefront/Domain/Cart/Demo/CartConverter.cs]: Same naming style as in other parts of code. platformDto, paymentDto, coreDto...

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs [hunk 18]
```diff
@@ -45,59 +44,57 @@ public class DemoCartService : CartService
...
+                        await FillProductPartsOfGroups(cart, language, currency);
 
+                        result.Add(cart);
+                    }
+                    return new StaticPagedList<ShoppingCart>(result, criteria.PageNumber, criteria.PageSize, resultDto.TotalCount.Value);
+                });
+            }
 
-        protected virtual async Task AddConfiguredItemsToCartAsync(ShoppingCart cart, Language language, Currency currency)
+            return IntertallSearchCartsAsync();
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs]: Remove local function

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs [hunk 19]
```diff
@@ -45,59 +44,57 @@ public class DemoCartService : CartService
             _catalogService = catalogService;
         }
 
-        public override async Task<IPagedList<ShoppingCart>> SearchCartsAsync(CartSearchCriteria criteria)
+        public override Task<IPagedList<ShoppingCart>> SearchCartsAsync(CartSearchCriteria criteria)
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs]: Async to end

File: VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs [hunk 20]
```diff
@@ -45,59 +44,57 @@ public class DemoCartService : CartService
...
+                    }
+                    return new StaticPagedList<ShoppingCart>(result, criteria.PageNumber, criteria.PageSize, resultDto.TotalCount.Value);
+                });
+            }
 
-        protected virtual async Task AddConfiguredItemsToCartAsync(ShoppingCart cart, Language language, Currency currency)
+            return IntertallSearchCartsAsync();
+        }
+
+        protected virtual async Task FillProductPartsOfGroups(ShoppingCart cart, Language language, Currency currency)
```
Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs]: Async methods should end with Async suffix

Comment (pushnitsa) [inline VirtoCommerce.Storefront/Domain/Cart/Demo/DemoCartService.cs]: Remove unused variables

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**